command/connect/envoy/envoy.go

```diff
@@ -233,7 +265,29 @@ func (c *cmd) Run(args []string) int {
 
 		var proxyConf *api.AgentServiceConnectProxyConfig
 
-		if lanAddr != "" {
+		if len(c.bindAddresses) > 0 {
+			// override all default binding rules and just bind to the user-supplied addresses
+			bindAddresses := make(map[string]api.ServiceAddress)
+
+			for addrName, addrStr := range c.bindAddresses {
+				addr, port, err := parseAddress(addrStr)
+				if err != nil {
+					c.UI.Error(fmt.Sprintf("Failed to parse the bind address: %s=%s: %v", addrName, addrStr, err))
+					return 1
+				}
+
+				bindAddresses[addrName] = api.ServiceAddress{Address: addr, Port: port}
+			}
+
+			proxyConf = &api.AgentServiceConnectProxyConfig{
+				Config: map[string]interface{}{
+					"envoy_mesh_gateway_no_default_bind": true,
+					"envoy_mesh_gateway_bind_addresses":  bindAddresses,
+				},
+			}
+		} else if canBind(lanAddr) && canBind(wanAddr) {
```

Comment:
what happens if this is also false and we can't bind to either? should the gateway registration fail?

Reply:
Right now it lets Consul determine what to do and doesn't override any behavior with proxy config. This will actually cause it to bind to the lanAddr though so you are right that we should fail the registration in one circumstance.

Reply:
I made it fail if `!canBind(lanAddr) && lanAddr != ""` to prevent putting invalid configuration into the services registration.